Update interpolate to use new upsample overloads

dreiss opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37177 Update interpolate to use new upsample overloads**
* #37176 Add interpolate-style overloads to aten::upsample* ops
* #37175 Add support for float[]? arguments in native_functions.yaml

- Use new overloads that better reflect the arguments to interpolate.
- More uniform interface for upsample ops allows simplifying the Python code.
- Also reorder overloads in native_functions.yaml to give them priority.


Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991/)

Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991)

dreiss committed:
Update interpolate (half-baked)



Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991/)

[ghstack-poisoned]

dreiss committed:
Update on "Update interpolate to use new upsample overloads"


- Use new overloads that better reflect the arguments to interpolate.
- More uniform interface for upsample ops allows simplifying the Python code.
- Also reorder overloads in native_functions.yaml to give them priority.


Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991/)

Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991)

[ghstack-poisoned]

dreiss commented:
@lara-hdr , it looks like this is breaking ONNX export.  My test program is

```
class Model(torch.nn.Module):
    def forward(self, arg):
        return torch.nn.functional.interpolate(arg, scale_factor=2, recompute_scale_factor=False)

print(torch.jit.trace(Model(), torch.zeros(1,1,1,1)).code)
```

This now prints

```
def forward(self,
    input: Tensor) -> Tensor:
  _0 = torch.upsample_nearest2d(input, None, [2., 2.])
  return _0
```

(much simpler than the old version), but running

```
print(torch.jit.trace(Model(), torch.zeros(1,1,1,1)).code)
```

results in an exception: https://gist.github.com/dreiss/33963e24f89a1ac8b8387d5f11bfc71a

Do you know what the proper fix is?

lara-hdr commented:
Thanks for the heads up @dreiss.
@neginraoof to update the export functions.

dreiss committed:
Update on "Update interpolate to use new upsample overloads"


- Use new overloads that better reflect the arguments to interpolate.
- More uniform interface for upsample ops allows simplifying the Python code.
- Also reorder overloads in native_functions.yaml to give them priority.


Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991/)

Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991)

[ghstack-poisoned]

dreiss committed:
Update on "Update interpolate to use new upsample overloads"


- Use new overloads that better reflect the arguments to interpolate.
- More uniform interface for upsample ops allows simplifying the Python code.
- Also reorder overloads in native_functions.yaml to give them priority.


Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991/)

Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991)

[ghstack-poisoned]

dreiss commented:
@neginraoof , any thoughts on this?

dreiss commented:
@lara-hdr, do you know if someone can look at the ONNX failures here?  I would love to find a way to land this.

lara-hdr commented:
@spandantiwari @neginraoof to look into the onnx failures

neginraoof commented:
Sorry I haven't got to build these changes locally and verify. I'm working on that.

From looking at the error at line: https://github.com/pytorch/pytorch/blob/461014d54b3981c8fa6617f90ff7b7df51ab1e85/torch/onnx/symbolic_helper.py#L283
Our ONNX symbolic argparser does not support float list. So I think it (https://github.com/pytorch/pytorch/blob/461014d54b3981c8fa6617f90ff7b7df51ab1e85/torch/onnx/symbolic_helper.py#L53) needs to be extended with a new arg descriptor, maybe 'fs', such how int list is handled.

dreiss commented:
Thanks, @neginraoof!

> But it's not clear to me what's the use-case of the other upsample function with schema:
upsample_nearest2d(Tensor self, int[2] output_size, float? scales_h=None, float? scales_w=None)

These are the overloads currently used in production.  After this PR lands, they will no longer be used by the Python frontend, and only kept for backward compatibility.

dreiss committed:
Update on "Update interpolate to use new upsample overloads"


- Use new overloads that better reflect the arguments to interpolate.
- More uniform interface for upsample ops allows simplifying the Python code.
- Also reorder overloads in native_functions.yaml to give them priority.


Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991/)

Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991)

[ghstack-poisoned]

dreiss committed:
Update on "Update interpolate to use new upsample overloads"


- Use new overloads that better reflect the arguments to interpolate.
- More uniform interface for upsample ops allows simplifying the Python code.
- Also reorder overloads in native_functions.yaml to give them priority.


Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991/)

Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991)

[ghstack-poisoned]

dreiss committed:
Update on "Update interpolate to use new upsample overloads"


- Use new overloads that better reflect the arguments to interpolate.
- More uniform interface for upsample ops allows simplifying the Python code.
- Also reorder overloads in native_functions.yaml to give them priority.


Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991/)

Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991)

[ghstack-poisoned]

dreiss committed:
Update on "Update interpolate to use new upsample overloads"


- Use new overloads that better reflect the arguments to interpolate.
- More uniform interface for upsample ops allows simplifying the Python code.
- Also reorder overloads in native_functions.yaml to give them priority.


Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991/)

Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991)

[ghstack-poisoned]

dreiss committed:
Update on "Update interpolate to use new upsample overloads"


- Use new overloads that better reflect the arguments to interpolate.
- More uniform interface for upsample ops allows simplifying the Python code.
- Also reorder overloads in native_functions.yaml to give them priority.


Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991/)

Differential Revision: [D21209991](https://our.internmc.facebook.com/intern/diff/D21209991)

[ghstack-poisoned]

gchanan commented on torch/csrc/jit/passes/quantization/quantization_patterns.h:
```diff
@@ -798,6 +798,29 @@ graph(%a_quant, %alpha, %scale, %input_scale, %r_scale, %r_zero_point, %r_dtype)
       "aten::mean", {"%dim", "%keepdim", "%out"});
 
+  auto upsample_nearest1d_vec = getInputTensorQParamOpFusionInfo(
+      "aten::upsample_nearest1d", {"%output_size", "%scale_factors"});
```
these aren't overload-name based?

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6058,6 +6058,93 @@
+    CPU: upsample_nearest3d_backward_cpu
+    CUDA: upsample_nearest3d_backward_cuda
+
 - func: upsample_linear1d.out(Tensor self, int[1] output_size, bool align_corners, float? scales=None, *, Tensor(a!) out) -> Tensor(a!)
```
can we comment somewhere that these only exist for serialized?

Note you should be able to kill them now by writing a serialization upgrader, see https://github.com/pytorch/pytorch/pull/40279.  I believe there's something similar for mobile.

I think just commenting is good enough for now, though.

gchanan replied:
Well actually we couldn't do that until we update the C++ api as well.

gchanan commented on torch/nn/functional.py:
```diff
@@ -3053,89 +3053,101 @@ def interpolate(input, size=None, scale_factor=None, mode='nearest', align_corne
-    if size is None and scale_factor is None:
-        raise ValueError('either size or scale_factor should be defined')
+    # Process size and scale_factor.  Validate that exactly one is set.
+    # Validate it's length if it is a list, or expand it if it is a scalar.
```
nit: it's -> its

gchanan commented on torch/nn/functional.py:
```diff
@@ -3053,89 +3053,101 @@ def interpolate(input, size=None, scale_factor=None, mode='nearest', align_corne
-                if is_float_scale_factor:
+                if math.floor(scale) != scale:
+                    warnings.warn("The default behavior for interpolate/upsample with float scale_factor will change "
+                                  "in 1.6.0 to align with other frameworks/libraries, and use scale_factor directly, "
```
Umm, according to the documentation above, didn't this already change?

gchanan commented on torch/nn/functional.py:
```diff
@@ -3053,89 +3053,101 @@ def interpolate(input, size=None, scale_factor=None, mode='nearest', align_corne
-            is_float_scale_factor = False
+    if recompute_scale_factor is None:
+        # TODO: Change default and update warning.
+        recompute_scale_factor = True
```
wait, isn't the default false now?
